include/red_black_tree.h

```diff
@@ -128,6 +141,231 @@ class RedBlackTree : public BinarySearchTree<K, V> {
+    //
+    // 这很好办，没有兄弟，创造兄弟，把 D 让给你 B* 了，对 C 做一个左旋操作。
+    //
+    //                C
+    //              /   \
+    //             A!    E
+    //           /  \
+    //          B*   D
+    //
+    void remove_fixup(Node* x, Node* xp) {
+        while (xp && (!x || x->color == ColorType::BLACK)) {
+            if (!x || x == xp->left) { // x && x== xp->left if using successor
+                auto w = xp->right;
+                if (w && w->color == ColorType::RED) { // case 4: 没兄弟，制造兄弟
+                    w->color = ColorType::BLACK;
+                    xp->color = ColorType::RED;
+                    left_rotate(xp);
+                    // 转向 case 1/2/3
+                    w = xp->right;
+                }
+
+                // 到这里，w 一定是黑色的
+                if (!w || (black(w->left) && black(w->right))) { // case 3: 兄弟也穷
+                    if (w) w->color = ColorType::RED;
+                    x = xp;
+                    xp = x->p;
+                } else {
+                    // 这里 w 一定不空，因为兄弟有钱啊
+                    if (black(w->right)) { // case 2: 兄弟有钱，但是藏起来了
+                        w->left->color = ColorType::BLACK;
+                        w->color = ColorType::RED;
+                        right_rotate(w);
+                        w = xp->right;
+                    }
+
+                    // case 1: 兄弟有钱，也愿意贡献
+                    w->right->color = ColorType::BLACK;
+                    w->color = xp->color;
+                    xp->color = ColorType::BLACK;
+                    left_rotate(xp);
+                    break;
+                }
+            } else {
+                // 镜像算法
+                auto w = xp->left;
+                if (w && w->color == ColorType::RED) { // case 4: 没兄弟，制造兄弟
+                    w->color = ColorType::BLACK;
+                    xp->color = ColorType::RED;
+                    right_rotate(xp);
+                    // 转向 case 1/2/3
+                    w = xp->left;
+                }
+
+                // 到这里，w 一定是黑色的
+                if (!w || (black(w->left) && black(w->right))) { // case 3: 兄弟也穷
+                    if (w) w->color = ColorType::RED;
+                    x = xp;
+                    xp = x->p;
+                } else {
+                    // 这里 w 一定不空
+                    if (black(w->left)) { // case 2: 兄弟有钱，但是藏起来了
+                        w->right->color = ColorType::BLACK;
+                        w->color = ColorType::RED;
+                        left_rotate(w);
+                        w = xp->left;
+                    }
+
+                    // case 1: 兄弟有钱，也愿意贡献
+                    w->left->color = ColorType::BLACK;
+                    w->color = xp->color;
+                    xp->color = ColorType::BLACK;
+                    right_rotate(xp);
+                    x = _root;
+                    break;
+                }
+            }
+        }
+
+        if (x) x->color = ColorType::BLACK;
+    }
+
+    bool black(Node* u) const {
```

Comment:
I'd like to make `black` as a member method of `Node`.

Reply:
No, null node is black.